### ISSUE TITLE
'prisma migrate dev' using url instead of shadowDatabaseUrl

### ISSUE THREAD

**Issue by isaacgr** [prisma/prisma#20690]
### Bug description

When running `prisma migrate dev` it uses `url` as opposed to `shadowDatabaseUrl`.

```
ubuntu@dev-environment:~/gitrepos/yyy$ npx prisma migrate dev --name init
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "yyy" at "xxxx:3306"

Error: P1010: User `xxx` was denied access on the database `mysql`
```

### How to reproduce

1. Set the shadowDatabaseUrl variable in the `schema.prisma` file to a different database than what's defined in the `url` variable
2. Run `npx prisma migrate dev`
3. The above error will be seen with the database defined in the `url` variable, indicating that its not using the one defined by the `shadowDatabaseUrl`

### Prisma information

```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider          = "mysql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
}
```

### Environment & setup

- OS: Linux
- Database: MariaDB
- Node.js version: v20.5.1

### Prisma Version

```
prisma                  : 5.1.1
@prisma/client          : 5.1.1
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.1.1-1.6a3747c37ff169c90047725a05a6ef02e32ac97e
Default Engines Hash    : 6a3747c37ff169c90047725a05a6ef02e32ac97e
Studio                  : 0.492.0
```

**Comment by janpio** [comment]
The `shadowDatabaseUrl` is only used for shadow database functionality. You are still trying to create a migration for your actual database defined in `url`, so it has to talk to your normal database. And if that does not work, then it can not proceed and get to the actual step where it would try to talk to the shadow database.

Do you agree that there is no problem then and this issue can be closed?

**Comment by isaacgr** [comment]
> The `shadowDatabaseUrl` is only used for shadow database functionality. You are still trying to create a migration for your actual database defined in `url`, so it has to talk to your normal database. And if that does not work, then it can not proceed and get to the actual step where it would try to talk to the shadow database.
> 
> Do you agree that there is no problem then and this issue can be closed?

The documentation (and what I gathered from the linked ticket) seems to imply that if I didn't want to give `*.*` permissions to my database user that I would instead use the shadow database url definition.

https://www.prisma.io/docs/concepts/components/prisma-migrate/shadow-database#shadow-database-user-permissions

So if that's not the case then how would I get around not giving those permissions to my database user? I only want them to act on the database defined in the databaseUrl. So only grant all those permissions on that single database.

**Comment by janpio** [comment]
You give the needed shadow database permissions to the user you use in `shadowDatabaseUrl`, and you use limited permissions for your normal database connection string in `url`. This way `migrate dev` can do shadow database things with that connection string, and normal "let me figure out what the state of the database is" with the normal connection string and potentially run the migrations later.

Note: `url` and `shadowDatabaseUrl` should point to _different_ databases. Usually Prisma just creates a temporary shadow database with the `url` credentials, but if you don't want to allow that, you have to give Prisma an already existing database to work with.

**Comment by isaacgr** [comment]
Right. They do point to different databases, and the grants for my user look like the below

```
| GRANT ALL PRIVILEGES ON `database`.* TO `dev`@`%`                                                         |
| GRANT ALL PRIVILEGES ON `database_dev`.* TO `dev`@`%`                                                     |
+----------------------------------------------------------------------------------------------------------------+
```

Where database is the one used by databaseUrl and database_dev is used by the shadow.

The problem is that `prisma migrate dev` still seems to require that I setup `*.*` permissions which I don't want to do.

**Comment by janpio** [comment]
Do you get the same error when running `npx prisma db pull` with your main database credentials?

**Comment by janpio** [comment]
Actually, we looked into the code a bit: `mysql` as database name is only used when the defined database name (form the connection string) does not exist, and Prisma needs to actually create the database first. In MySQL you have to connect to _a_ database, and `mysql` as a system database usually exists.

Is it possible that `yyy` (from your output) does not actually exist yet?
Does the problem go away if you create that database manually first? Then Prisma should not try to connect to `mysql` at all.

**Comment by isaacgr** [comment]
Ya that's the thing. Both of those databases already exist and my user has those above permissions on them. I am using mariadb if that matters.

**Comment by janpio** [comment]
It shouldn't. 

Can you recount how you created and configured that user? I guess that is what we would need to have to reproduce the problem you describe.

**Comment by isaacgr** [comment]
1. `create user 'dev' identified by 'password'`
2. `grant all privileges on database.* to 'dev'@'%';`
3. `grant all privileges on database_dev.* to 'dev'@'%';`

This is what I did in AWS to create the user and grant permissions on the pre-made databases. I am also connecting via an ssl certificate if that's relevant.

**Comment by janpio** [comment]
Thank you, with that someone should be able to reproduce your setup (2 databases `database` and `database_dev`, one for `url` and other for `shadowDatabaseUrl`) on our side and see if they get the same error you are getting.
